src/data/markdown/translated-guides/en/07 Misc/05 k6 Extensions.md

```diff
@@ -1,4 +1,393 @@
+
+```go
+package log
+
+import (
+	"fmt"
+	"io"
+
+	"go.k6.io/k6/output"
+	"go.k6.io/k6/stats"
+)
+
+// Register the extension on module initialization.
+func init() {
+	output.RegisterExtension("logger", New)
+}
+
+// Logger writes k6 metric samples to stdout.
+type Logger struct {
+	out io.Writer
+}
+
+// New returns a new instance of Logger.
+func New(params output.Params) (output.Output, error) {
+	return &Logger{params.StdOut}, nil
+}
+
+// Description returns a short human-readable description of the output.
+func (*Logger) Description() string {
+	return "logger"
+}
+
+// Start initializes any state needed for the output, establishes network
+// connections, etc.
+func (o *Logger) Start() error {
+	return nil
+}
+
+// AddMetricSamples receives metric samples from the k6 Engine as they're
+// emitted and prints them to stdout.
+func (l *Logger) AddMetricSamples(samples []stats.SampleContainer) {
+	for i := range samples {
+		all := samples[i].GetSamples()
+		for j := range all {
+			fmt.Fprintf(l.out, "%d %s: %f\n", all[j].Time.UnixNano(), all[j].Metric.Name, all[j].Value)
+		}
+	}
+}
+
+// Stop finalizes any tasks in progress, closes network connections, etc.
+func (*Logger) Stop() error {
+	return nil
+}
+```
+
+</CodeGroup>
+
+Notice a couple of things:
+
+- The module initializer `New()` receives an instance of
+  [`output.Params`](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L36).
+  With this object the extension can access the output-specific configuration,
+  interfaces to the filesystem, synchronized stdout and stderr, and more.
+- `AddMetricSamples` in this example simply writes to stdout. In a real-world
+  scenario this output might have to be buffered and flushed periodically to avoid
+  memory leaks. Below we'll discuss some helpers you can use for that.
+
+
+#### Additional features
+
+- Output structs can optionally implement additional interfaces that allows them to
+  [receive thresholds](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L79),
+  [test run status updates](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L94)
+  or [interrupt a test run](https://github.com/grafana/k6/blob/v0.33.0/output/types.go#L88).
+- Because output implementations typically need to process large amounts of data that
+  k6 produces and dispatch it to another system, we've provided a couple of helper
+  structs you can use in your extensions:
+  [`output.SampleBuffer`](https://github.com/grafana/k6/blob/v0.33.0/output/helpers.go#L35)
+  is a thread-safe buffer for metric samples to help with memory management and
+  [`output.PeriodicFlusher`](https://github.com/grafana/k6/blob/v0.33.0/output/helpers.go#L75)
+  will periodically run a function which is useful for flushing or dispatching the
+  buffered samples.
+  For usage examples see the [`statsd` output](https://github.com/k6io/k6/blob/v0.33.0/output/statsd/output.go#L55).
+
+
+#### Things to keep in mind
+
+- ...
```

Comment:
Any ideas what to mention here? I considered memory management and no blocking, but it's mentioned above.

Should we drop or rename this section?

Reply:
Couldn't think of anything, so decided to drop it.